vermouth/tests/test_graph_utils.py

```diff
@@ -0,0 +1,384 @@
+      {'id': [0, 1]},
+      [{'id': 0}, {'id': 1}], {(0, 1): {'weight': 1.5}}),
+     ([{}, {}, {}, {}], {(0, 1): {}, (1, 2): {}, (2, 3): {}}, [[0, 1], [2, 3]],
+      {'id': [0, 1], 'attr': ['a', 'b']},
+      [{'id': 0, 'attr': 'a'}, {'id': 1, 'attr': 'b'}], {(0, 1): {'weight': 1.0}}),
+     pytest.param([{}, {}, {}, {}], {(0, 1): {}, (1, 2): {}, (2, 3): {}}, [[0, 1], [2, 3]],
+      {'id': [1], 'attr': ['a', 'b']},
+      [{'id': 0, 'attr': 'a'}, {'id': 1, 'attr': 'b'}], {(0, 1): {'weight': 1.0}},
+      marks=pytest.mark.xfail(raises=IndexError)),
+
+    ]
+   )
+def test_blockmodel(node_data, edges, partitions, attrs, expected_nodes, expected_edges):
+    graph = basic_molecule(node_data, edges)
+    found = vermouth.graph_utils.blockmodel(graph, partitions, **attrs)
+    expected = basic_molecule(expected_nodes, expected_edges)
+    pprint(found.nodes(data=True))
+    pprint(expected.nodes(data=True))
+
+    for node in found:
+        data = found.nodes[node]
+        subgraph = data['graph']
+        assert len(subgraph.nodes) == data['nnodes']
+        assert len(subgraph.edges) == data['nedges']
+        assert nx.density(subgraph) == data['density']
+        del found.nodes[node]['graph']
+        del found.nodes[node]['nnodes']
+        del found.nodes[node]['nedges']
+        del found.nodes[node]['density']
+
+    assert found.nodes(data=True) == expected.nodes(data=True)
+    edges_seen = set()
+    for idx, jdx, data in found.edges(data=True):
+        assert expected.has_edge(idx, jdx) and expected.edges[idx, jdx] == data
+        edges_seen.add(frozenset((idx, jdx)))
+    assert set(frozenset(edge) for edge in expected.edges) == edges_seen
+
+
+def test_blockmodel_graph_attr():
+    graph = basic_molecule([{}, {}, {}], {(0, 1): {}, (1, 2): {}})
+    found = vermouth.graph_utils.blockmodel(graph, [[0], [1, 2]])
+    assert found.nodes[0]['nnodes'] == len(found.nodes[0]['graph'].nodes) == 1
+    assert found.nodes[1]['nnodes'] == len(found.nodes[1]['graph'].nodes) == 2
+    assert found.nodes[0]['nedges'] == len(found.nodes[0]['graph'].edges) == 0
+    assert found.nodes[1]['nedges'] == len(found.nodes[1]['graph'].edges) == 1
+
+
+@pytest.mark.parametrize('nodes1, nodes2, match, expected',
+        [
+         ([], [], {}, 0),
+         ([{'atomname': 0}], [{'atomname': 1}], {0:0}, 0),
+         ([{'atomname': 0}], [{'atomname': 1}], {}, 0),
+         ([{'atomname': 0}], [{'atomname': 0}], {}, 0),
+         ([{'atomname': 0}], [{'atomname': 0}], {0:0}, 1),
+         ([{'atomname': 0}, {'atomname': 0}], [{'atomname': 0}, {'atomname': 0}], {0:0}, 1),
+         ([{'atomname': 0}, {'atomname': 0}], [{'atomname': 0}, {'atomname': 0}], {0:0, 1:1}, 2),
+         ([{'atomname': 0}, {'atomname': 0}], [{'atomname': 0}, {'atomname': 1}], {0:0, 1:1}, 1),
+         ([{'atomname': 0}, {'atomname': 1}], [{'atomname': 0}, {'atomname': 1}], {0:0, 1:1}, 2),
+        ]
+        )
+def test_rate_match(nodes1, nodes2, match, expected):
```

Comment:
The `match` parameter of `rate_match` is not documented. Also, could `match` be `{0:1}`?